src/fsharp/FSharp.Core/control.fs

```diff
@@ -1178,14 +1222,29 @@ namespace Microsoft.FSharp.Control
     // Contains helpers that will attach continuation to the given task.
     // Should be invoked as a part of protectedPrimitive(withResync) call
     module TaskHelpers = 
+        // This uses a trick to get the underlying OperationCanceledException
+        let inline getCancelledException (completedTask:Task) (waitWithAwaiter) =
+            let fallback = new TaskCanceledException(completedTask) :> OperationCanceledException
```

Comment:
Why allocate the fallback here when it isn't needed on all paths?

Reply:
Ok I guess I make this a function instead.